### PR TITLE
Fixed: Alias dropped when new lot of compound gets registered in bulk loader

### PR DESCRIPTION
**Pull request by dalejerikson**
**Steps to Reproduce**

Bulk load a new structure with a Parent Alias mapped in. This creates a new parent with the alias, and a new lot.

Then load the same structure with a different Parent Alias value. This will create a new lot on the existing parent.

**Bug Behavior**

A new lot is created, but the alias from the second upload is ignored. The parent ends up with one alias (only the one from the first upload).

**Fixed Behavior**

A new lot will be created and the existing parent will be updated to add the new Parent Alias value. The parent will end up with two aliases (one from each upload)

**Example**

One of the tests I conducted was to have one compound with the alias "Dale's Favorite Compound" and I uploaded the same structure with the alias of "Second Alias". The result was two lots with the same parent. The parent contains both aliases. 

Screen Shot 2022-11-02 at 11 09 04 AM

### CODE REVIEW

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java [hunk 1]
```diff
@@ -825,9 +825,77 @@ public Parent validateParent(Parent parent, Collection<BulkLoadPropertyMappingDT
...
+						// however due to readability and the assumingly small number of aliases 
+						// parents will have this is fine 
+
+						Set<String> parentAliasStrings = new HashSet<>();
+						for(ParentAlias alias : parent.getParentAliases()){
+							parentAliasStrings.add(alias.toString());
+						}
+						Set<String> foundParentAliasStrings = new HashSet<>(); 
+						for(ParentAlias alias : foundParent.getParentAliases()){
+							foundParentAliasStrings.add(alias.toString());
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:844]: What does this `alias.toString()` method return? Is it just the `aliasName`? Or is it an Object like `ParentAlias[lsType=null,lsKind=null,id=27,aliasName="my favorite alias"]`?

**Comment by bffrost** [comment]
@dalejerikson can you add some acasclient tests for this new behavior? Specifically some confirming the "negative" cases of loading the same alias or aliases twice and confirming we don't create duplicate aliases.

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java [hunk 2]
```diff
@@ -825,9 +825,77 @@ public Parent validateParent(Parent parent, Collection<BulkLoadPropertyMappingDT
 							|| foundParent.getCorpName().contains(parent.getLabelPrefix().getLabelPrefix()));
 					if (sameStereoCategory & sameStereoComment
 							& (sameCorpName | (noCorpName & sameCorpPrefixOrNoPrefix))) {
-						// parents match
-						parent = foundParent;
-						break searchResultLoop;
+						// parents match (based on above criteria)
+						boolean equalAliases = true; // assumption is aliases are same unless proven to be false
+
+						// Need to Create String Sets to Do
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:831]: nit: Just because a bunch of the functions in this `BulkLoadServiceImpl` are already quite long, It'd be nice to encapsulate this into a helper function like `private Parent checkAndUpdateParentAliases(Parent parent, Parent foundParent)`. That'll help maintain readability in this already long `validateParent` function.

**Review by bffrost — CHANGES_REQUESTED** [review]
@dalejerikson the general flow of the logic looks good, but I'm concerned that the `toString()` method may not work the way you expect it does. The best way to confirm this would be to write a few acasclient tests and post them for PR.

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java (hunk 1 above)
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:844]: I suppose `getAlias()` would be more appropriate. Also I forgot to add in if it is marked as ignored so I will do that as well.

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java (hunk 2 above)
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:831]: That's fair.

**Comment by dalejerikson** [comment]
Interesting: When adding tests to acas client, I noticed that there is already something that tests manipulating parent aliases.

https://github.com/mcneilco/acasclient/blob/release/2022.4.x/tests/test_acasclient.py#L3081 

Before this fix, it theoretically should have failed. Looks like the logic it different and more directly interacts with the Parent Alias object than the bug resulting from the issue in the bulk loader that this ticket fixes. 

Will request re-review for this now but i'll look to add tests that more accurately reflect this ticket.

**Comment by dalejerikson** [comment]
ACAS Client Test PR: https://github.com/mcneilco/acasclient/pull/101

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java (hunk 2 above)
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:831]: @bffrost Comment was marked as resolved but I don't think it actually was, right?

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java [hunk 3]
```diff
@@ -825,9 +825,83 @@ public Parent validateParent(Parent parent, Collection<BulkLoadPropertyMappingDT
...
+						// "Is alias (str) of parent in aliases of foundParent (set of strs)?" (and vice versa) logic 
+
+						// Note: This is not the most efficient / concise code to do this process
+						// however due to readability and the assumingly small number of aliases 
+						// parents will have this is fine 
+
+						Set<String> parentAliasStrings = new HashSet<>();
+						for(ParentAlias alias : parent.getParentAliases()){
+							if (! alias.isIgnored())
+							{
```
Comment (brianbolt) [inline src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:841]: nit - Can we stick with { being on the same line as the if or else block since it's that way in most of the code?

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java [hunk 4]
```diff
@@ -825,9 +825,83 @@ public Parent validateParent(Parent parent, Collection<BulkLoadPropertyMappingDT
...
+
+						// If Incoming Parent and Found Parent Have Equal Aliases 
+						if(equalAliases)
+						{
+							// No Changes or Modifications Needed; Can Break Loop
+							// Continue 
+							parent = foundParent;
+							break searchResultLoop; 
+						}
+						else
```
Comment (brianbolt) [inline src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:884]: nit - Same nit as above with keeping else on the same line as } and {

**Review by bffrost — APPROVED** [review]
LGTM! The fixes for string comparison look good.

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java (hunk 2 above)
Comment (bffrost) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:831]: Good catch @brianbolt . @dalejerikson  Can you look at pulling this logic out into a helper function just to improve readability?

Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:831]: I left an additional separate comment about the break statement that is in the code

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java [hunk 5]
```diff
@@ -825,9 +825,75 @@ public Parent validateParent(Parent parent, Collection<BulkLoadPropertyMappingDT
...
+								}
+							}
+						}	
+
+						// If Incoming Parent and Found Parent Have Equal Aliases 
+						if(equalAliases) {
+							// No Changes or Modifications Needed; Can Break Loop
+							// Continue 
+							parent = foundParent;
+							break searchResultLoop;
```
[the remark this answers is not in the document]
Comment (bffrost) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:878]: What about two helpers? `compareParentAliases` and `updateParentAliases`
Then you can keep the `if(equalAliases)` block and the `break` statement, but still keep the complexity down.

[the remark this answers is not in the document]
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java:835]: Implemented

File: src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java (hunk 2 above)
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java]: Implemented
